hooks/tk-multi-publish2/basic/publish_session.py

```diff
@@ -0,0 +1,712 @@
+        # etc.
+        path = sgtk.util.ShotgunPath.normalize(path)
+
+        # determine the publish path, version, type, and name
+        publish_info = self._get_publish_info(path, settings)
+
+        publish_name = publish_info["publish_name"]
+
+        # see if there are any other publishes of this path with a status.
+        # Note the name, context, and path *must* match the values supplied to
+        # register_publish in the publish phase in order for this to return an
+        # accurate list of previous publishes of this file.
+        publishes = publisher.util.get_conflicting_publishes(
+            item.context,
+            path,
+            publish_name,
+            filters=["sg_status_list", "is_not", None]
+        )
+
+        if publishes:
+            conflict_info = (
+                "If you continue, these conflicting publishes will no longer "
+                "be available to other users via the loader:<br>"
+                "<pre>%s</pre>" % (pprint.pformat(publishes),)
+            )
+            self.logger.warn(
+                "Found %s conflicting publishes in Shotgun" %
+                (len(publishes),),
+                extra={
+                    "action_show_more_info": {
+                        "label": "Show Conflicts",
+                        "tooltip": "Show the conflicting publishes in Shotgun",
+                        "text": conflict_info
+                    }
+                }
+            )
+
+        # check to see if the next version of the work file already exists on
+        # disk. if so, warn the user and provide the ability to jump to save
+        # to that version now
+        (next_version_path, version) = self._get_next_version_info(
+            path, settings)
+        if next_version_path and os.path.exists(next_version_path):
+
+            # determine the next available version_number. just keep asking for
+            # the next one until we get one that doesn't exist.
+            while os.path.exists(next_version_path):
+                (next_version_path, version) = self._get_next_version_info(
+                    next_version_path, settings)
+
+            self.logger.error(
+                "The next version of this file already exists on disk.",
+                extra={
+                    "action_button": {
+                        "label": "Save to v%s" % (version,),
+                        "tooltip": "Save to the next available version number, "
+                                   "v%s" % (version,),
+                        "callback": lambda: _save_session(next_version_path)
+                    }
+                }
+            )
+            return False
+
+        # store the session path since this is guaranteed to run just before
+        # the publish itself
+        item.properties["work_file_path"] = path
+
+        # store the publish info in the properties
+        item.properties["publish_info"] = publish_info
```

Comment:
caching the calculated publish info here to reduce duplication in the `publish` method.